presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java

```diff
@@ -0,0 +1,307 @@
+
+                // cast expression if coercion is registered
+                Type coerceToType = expressionAnalysis.getCoercion(node);
+
+                if (coerceToType != null) {
+                    rewrittenExpression = new Cast(rewrittenExpression, coerceToType.getTypeSignature().toString());
+                }
+
+                return rewrittenExpression;
+            }
+        }, expression);
+
+        // replace qualified names with input references since row expressions do not support these
+        Expression expressionWithInputReferences = ExpressionTreeRewriter.rewriteWith(new SymbolToInputRewriter(symbolInput), coercedExpression);
+
+        // determine the type of every expression
+        IdentityHashMap<Expression, Type> expressionTypes = getExpressionTypesFromInput(
+                session,
+                metadata,
+                sqlParser,
+                inputTypes,
+                expressionWithInputReferences);
+
+        // convert to row expression
+        return translate(expressionWithInputReferences, SCALAR, expressionTypes, metadata.getFunctionRegistry(), metadata.getTypeManager(), session, true);
+    }
+
+    private static class CanonicalizationVisitor
+            implements RowExpressionVisitor<Void, RowExpression>
+    {
+        @Override
+        public RowExpression visitCall(CallExpression call, Void context)
+        {
+            call = new CallExpression(
+                    call.getSignature(),
+                    call.getType(),
+                    call.getArguments().stream()
+                            .map(expression -> expression.accept(this, context))
+                            .collect(toImmutableList()));
+
+            String callName = call.getSignature().getName();
+
+            if (callName.equals("AND") || callName.equals("OR")) {
+                // return sorted distinct arguments
+                Set<RowExpression> distinctArguments = ImmutableSet.copyOf(call.getArguments());
+                if (distinctArguments.size() == 1) {
+                    return Iterables.getOnlyElement(distinctArguments);
+                }
+
+                return new CallExpression(
+                        call.getSignature(),
+                        call.getType(),
+                        ROW_EXPRESSION_ORDERING.sortedCopy(distinctArguments));
+            }
+
+            if (callName.equals(mangleOperatorName(EQUAL)) || callName.equals(mangleOperatorName(NOT_EQUAL)) || callName.equals("IS_DISTINCT_FROM")) {
+                // sort arguments
+                return new CallExpression(
+                        call.getSignature(),
+                        call.getType(),
+                        ROW_EXPRESSION_ORDERING.sortedCopy(call.getArguments()));
+            }
+
+            if (callName.equals(mangleOperatorName(GREATER_THAN)) || callName.equals(mangleOperatorName(GREATER_THAN_OR_EQUAL))) {
```

Comment:
We are canonicalizing `>` to `<` and `>=` to `<=` by flipping the expressions, so we don't need anything special for LESS_THAN or LESS_THAN_OR_EQUAL as they are already "canonical".